activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb

```diff
@@ -364,6 +364,8 @@ def transaction_open?
       end
 
       def reset_transaction(restore: false) # :nodoc:
+        @transaction_manager.suspend_transactions if defined?(@transaction_manager)
```

Comment:
Why we need to use `defined?` here and not in the line below?

Reply:
Good question - since this method is called in `initialize` without the `restore` option, the `@transaction_manager` was always initialized before being accessed. Maybe it would be better to do something like this here:

```ruby
if restore
  @transaction_manager.suspend_transactions
  old_state = @transaction_manager if @transaction_manager.restorable?
end
```

but I'm going to rework this based on the additional feedback, so I'll remove this line entirely.